target/linux/ipq807x/files/arch/arm64/boot/dts/qcom/ipq8070-cax1800.dts

```diff
@@ -0,0 +1,514 @@
+// SPDX-License-Identifier: GPL-2.0-or-later OR MIT
+/* Copyright (c) 2021, Dirk Buchwalder <buchwalder@posteo.de> */
+
+/dts-v1/;
+
+#include "ipq8074.dtsi"
+#include "ipq8074-ac-cpu.dtsi"
+#include "ipq8074-ac-nss.dtsi"
+#include <dt-bindings/gpio/gpio.h>
+#include <dt-bindings/input/input.h>
+
+/ {
+	#address-cells = <2>;
+	#size-cells = <2>;
+
+	model = "Edimax CAX1800";
+	compatible = "edimax,cax1800", "qcom,ipq8074";
+	interrupt-parent = <&intc>;
+
+	aliases {
+		serial0 = &blsp1_uart5;
+		led-boot = &led_system_red;
+		led-failsafe = &led_system_red;
+		led-running = &led_system_green;
+		led-upgrade = &led_system_red;
+		/* Aliases as required by u-boot to patch MAC addresses */
+		ethernet4 = &dp5;
+		label-mac-device = &dp5;
+	};
+
+	chosen {
+		stdout-path = "serial0:115200n8";
+		bootargs-append = " root=/dev/ubiblock0_1";
+	};
+
+	keys {
+		compatible = "gpio-keys";
+
+		reset {
+			label = "reset";
+			gpios = <&tlmm 32 GPIO_ACTIVE_LOW>;
+			linux,code = <KEY_RESTART>;
+		};
+	};
+
+	leds {
+		compatible = "gpio-leds";
+
+		led_system_red: system-red {
+			label = "red:system";
+			gpios = <&tlmm 25 GPIO_ACTIVE_HIGH>;
+		};
+
+		led_system_green: system-green {
+			label = "green:system";
+			gpios = <&tlmm 26 GPIO_ACTIVE_HIGH>;
+		};
+
+		led_system_blue: system-blue {
+			label = "blue:system";
+			gpios = <&tlmm 27 GPIO_ACTIVE_HIGH>;
+		};
+	};
+};
+
+&tlmm {
+	mdio_pins: mdio-pins {
+		mdc {
+			pins = "gpio68";
+			function = "mdc";
+			drive-strength = <8>;
+			bias-pull-up;
+		};
+
+		mdio {
+			pins = "gpio69";
+			function = "mdio";
+			drive-strength = <8>;
+			bias-pull-up;
+		};
+	};
+};
+
+&blsp1_uart5 {
+	status = "okay";
+};
+
+&prng {
+	status = "okay";
+};
+
+&cryptobam {
+	status = "okay";
+};
+
+&crypto {
+	status = "okay";
+};
+
+&qpic_bam {
+	status = "okay";
+};
+
+&qpic_nand {
+	status = "okay";
+
+	nand@0 {
+		reg = <0>;
+		nand-ecc-strength = <4>;
+		nand-ecc-step-size = <512>;
+		nand-bus-width = <8>;
+
+		partitions {
+			compatible = "fixed-partitions";
+			#address-cells = <1>;
+			#size-cells = <1>;
+
+			partition@0 {
+				label = "rootfs";
+				reg = <0x0000000 0x3400000>;
+			};
+		};
+	};
+};
+
+&blsp1_spi1 {
+	pinctrl-0 = <&spi_0_pins>;
+	pinctrl-names = "default";
+	cs-select = <0>;
+	status = "ok";
+
+	m25p80@0 {
+		#address-cells = <1>;
+		#size-cells = <1>;
+		reg = <0>;
+		compatible = "jedec,spi-nor";
+		spi-max-frequency = <50000000>;
+		use-default-sizes;
+
+		partitions {
+			compatible = "fixed-partitions";
+			#address-cells = <1>;
+			#size-cells = <1>;
+
+			partition@480000 {
```

Comment:
Hm, there has gotta be more partitions for sure.
Also, can the SMEM parser be used for this?

Reply:
Hi Robert,
I've tried smem, but since the device is using NAND and NOR, this always led to overlapping partitions.

This is the smem output from uboot (looks like the standard smem layout):

`No.: Name             Attributes            Start             Size

  0: 0:SBL1           0x0000ffff              0x0          0x50000

  1: 0:MIBIB          0x001040ff          0x50000          0x10000

  2: 0:BOOTCONFIG     0x001040ff          0x60000          0x20000

  3: 0:BOOTCONFIG1    0x001040ff          0x80000          0x20000

  4: 0:QSEE           0x0000ffff          0xa0000         0x180000

  5: 0:QSEE_1         0x0000ffff         0x220000         0x180000

  6: 0:DEVCFG         0x0000ffff         0x3a0000          0x10000

  7: 0:DEVCFG_1       0x0000ffff         0x3b0000          0x10000

  8: 0:APDP           0x0000ffff         0x3c0000          0x10000

  9: 0:APDP_1         0x0000ffff         0x3d0000          0x10000

 10: 0:RPM            0x0000ffff         0x3e0000          0x40000

 11: 0:RPM_1          0x0000ffff         0x420000          0x40000

 12: 0:CDT            0x0000ffff         0x460000          0x10000

 13: 0:CDT_1          0x0000ffff         0x470000          0x10000

 14: 0:APPSBLENV      0x0000ffff         0x480000          0x10000

 15: 0:APPSBL         0x0000ffff         0x490000          0xa0000

 16: 0:APPSBL_1       0x0000ffff         0x530000          0xa0000

 17: 0:ART            0x0000ffff         0x5d0000          0x40000

 18: 0:ETHPHYFW       0x0000ffff         0x610000          0x80000

 19: rootfs           0x0100ffff              0x0        0x3400000

 20: 0:WIFIFW         0x0100ffff        0x3400000         0x800000

 21: rootfs_1         0x0100ffff        0x3c00000        0x3400000

 22: 0:WIFIFW_1       0x0100ffff        0x7000000         0x800000

`
Partitions 0 and 19 are overlapping due to the same start address.

This are the partition maps of the stock firmware, as you can see,at least NAND partitions differs from smem:

`Creating 20 MTD partitions on "spi32766.0":
[    2.432598] 0x000000000000-0x000000050000 : "0:SBL1"
[    2.438653] 0x000000050000-0x000000060000 : "0:MIBIB"
[    2.443766] 0x000000060000-0x000000080000 : "0:BOOTCONFIG"
[    2.448776] 0x000000080000-0x0000000a0000 : "0:BOOTCONFIG1"
[    2.454131] 0x0000000a0000-0x000000220000 : "0:QSEE"
[    2.459621] 0x000000220000-0x0000003a0000 : "0:QSEE_1"
[    2.464824] 0x0000003a0000-0x0000003b0000 : "0:DEVCFG"
[    2.469757] 0x0000003b0000-0x0000003c0000 : "0:DEVCFG_1"
[    2.474807] 0x0000003c0000-0x0000003d0000 : "0:APDP"
[    2.480305] 0x0000003d0000-0x0000003e0000 : "0:APDP_1"
[    2.485286] 0x0000003e0000-0x000000420000 : "0:RPM"
[    2.490216] 0x000000420000-0x000000460000 : "0:RPM_1"
[    2.495001] 0x000000460000-0x000000470000 : "0:CDT"
[    2.500264] 0x000000470000-0x000000480000 : "0:CDT_1"
[    2.504923] 0x000000480000-0x000000490000 : "0:APPSBLENV"
[    2.510169] 0x000000490000-0x000000530000 : "0:APPSBL"
[    2.515467] 0x000000530000-0x0000005d0000 : "0:APPSBL_1"
[    2.520462] 0x0000005d0000-0x000000610000 : "0:ART"
[    2.525981] 0x000000610000-0x000000690000 : "0:ETHPHYFW"
[    2.530569] 0x000000000000-0x000000800000 : "ALL_SPI"
[    2.537111] QPIC controller hw version Major:1, Minor:5
[    2.540625] nand: device found, Manufacturer ID: 0xc2, Chip ID: 0xa1
[    2.545377] nand: ONFI 10-Compliant Macronix MX30UF1G18AC
[    2.551954] nand: 128 MiB, SLC, erase size: 128 KiB, page size: 2048, OOB size: 64
[    2.557275] Creating 6 MTD partitions on "qcom_nand.0":
[    2.564715] 0x000000000000-0x000003400000 : "rootfs"
[    2.609761] mtd: device 20 (rootfs) set to be root filesystem
[    2.610025] mtdsplit: no squashfs found in "rootfs"
[    2.614484] 0x000003400000-0x000003c00000 : "0:WIFIFW"
[    2.626027] 0x000003c00000-0x000004600000 : "storage_ubi"
[    2.634146] 0x000004600000-0x000005000000 : "manufacture_ubi"
[    2.642505] 0x000005000000-0x000005a00000 : "nms_data_ubi"
[    2.650794] 0x000000000000-0x000008000000 : "ALL_NAND"`

So I decided to hard-code the needed partitions in the dts.

Reply:
Ok, it's a mess they made by not having a generic SMEM partition type for NAND and NOR but rather they made it platform-specific and then this happens.
Please, add all of the SPI-NOR partitions to the DTS.

Reply:
done